### PR TITLE
[SYCL] Add tests for SYCL 2020 function objects

### PR DESCRIPTION
**Pull request by maximdimakov**
Some tests are rewritten with SYCL2020 group algorithm functions and SYCL2020 function objects

### CODE REVIEW

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+#include <limits>
+#include <numeric>
+#include <vector>
+using namespace sycl;
+
+template <class SpecializationKernelName, int TestNumber>
+class exclusive_scan_kernel;
+
+// std::exclusive_scan isn't implemented yet, so use serial implementation
+// instead
```
Comment (Pennycook) [inline SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:22]: Now that we compile everything with C++17, can we replace this with `std::exclusive_scan`?

File: SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+#include <limits>
+#include <numeric>
+#include <vector>
+using namespace sycl;
+
+template <class SpecializationKernelName, int TestNumber>
+class inclusive_scan_kernel;
+
+// std::inclusive_scan isn't implemented yet, so use serial implementation
+// instead
```
Comment (Pennycook) [inline SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp:22]: Same question as above.

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 1 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:22]: I changed using of emu::exclusive scan to std::exclusive_scan

File: SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp (hunk 2 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp:22]: Changed

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+          typename OutputContainer::value_type identity) {
+  typedef typename InputContainer::value_type InputT;
+  typedef typename OutputContainer::value_type OutputT;
+  typedef class exclusive_scan_kernel<SpecializationKernelName, 0> kernel_name0;
+  typedef class exclusive_scan_kernel<SpecializationKernelName, 1> kernel_name1;
+  typedef class exclusive_scan_kernel<SpecializationKernelName, 2> kernel_name2;
+  typedef class exclusive_scan_kernel<SpecializationKernelName, 3> kernel_name3;
+  OutputT init = 42;
+  size_t N = input.size();
+  size_t G = 64;
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: Why G is 64? Can you please move it to the place in the code where it is used? Or pass it as a reference to the `test` func. Make it `constexpr`.

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+void test(queue q, InputContainer input, OutputContainer output,
+          BinaryOperation binary_op,
+          typename OutputContainer::value_type identity) {
+  typedef typename InputContainer::value_type InputT;
+  typedef typename OutputContainer::value_type OutputT;
+  typedef class exclusive_scan_kernel<SpecializationKernelName, 0> kernel_name0;
+  typedef class exclusive_scan_kernel<SpecializationKernelName, 1> kernel_name1;
+  typedef class exclusive_scan_kernel<SpecializationKernelName, 2> kernel_name2;
+  typedef class exclusive_scan_kernel<SpecializationKernelName, 3> kernel_name3;
+  OutputT init = 42;
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: Please move it to the place in the code where it is used

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  size_t N = input.size();
+  size_t G = 64;
+  std::vector<OutputT> expected(N);
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:49]: Move it under 55 line

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  {
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:54]: Small comments regarding what checks do will be useful, e.g. for this one:
```
// checking 
// template <typename Group, typename T, typename BinaryOperation>
// T exclusive_scan_over_group(Group g, T x, BinaryOperation binary_op);
```

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                                 std::numeric_limits<int>::lowest());
+
+  test<class KernelNamePlusI>(q, input, output, sycl::plus<int>(), 0);
+  test<class KernelNameMinimumI>(q, input, output, sycl::minimum<int>(),
+                                 std::numeric_limits<int>::max());
+  test<class KernelNameMaximumI>(q, input, output, sycl::maximum<int>(),
+                                 std::numeric_limits<int>::lowest());
+
+#ifdef SPIRV_1_3
+  test<class KernelName_VzAPutpBRRJrQPB>(q, input, output,
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: Are there no inconsistency between kernel names in SPIR-V and kernel names here from one compilation to another? Why they look like this?

File: SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -I . -o %t.out
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp:2]: Please apply all comments from `exclusive_scan_sycl2020.cpp` to this test too.

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    return 0;
+  }
+
+  constexpr int N = 128;
+  std::array<int, N> input;
+  std::array<int, N> output;
+  std::iota(input.begin(), input.end(), 0);
+  std::fill(output.begin(), output.end(), 0);
+
+  test<class KernelNamePlusV>(q, input, output, sycl::plus<>(), 0);
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:159]: Does `V` for `void`?

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 3 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: It is used in the handler::parallel_for nd_range argument slighty lower. These tests are just copy of existing tests but with using of sycl2020 functions and function objects. I didn't change their work

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 7 above)
Comment (dm-vodopyanov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: @AlexeySotkin, @AlexeySachkov, can you please explain this if you know?

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 4 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: Moved

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 5 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:49]: Moved "init" variable. The others are used slighty lower in the test

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 6 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:54]: Added comments

File: SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp (hunk 8 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp:2]: Changes from exclusive_scan_sycl2020.cpp applied to inclusive_scan_sycl2020.cpp

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 9 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:159]: For empty template argument

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 7 above)
Comment (AlexeySachkov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: > Are there no inconsistency between kernel names in SPIR-V and kernel names here from one compilation to another?

Could you please explain in more details what do you mean?

> Why they look like this?

It is up to user how to name kernels. Probably those are just placeholders until better name is found?

Comment (dm-vodopyanov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: @AlexeySachkov, thanks! @maximdimakov, can you please rename them in this case?

Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: I renamed them in consistency with another kernel names

File: SYCL/GroupAlgorithm/reduce_sycl2020.cpp [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+                                 std::numeric_limits<int>::lowest());
+
+  test<class KernelNamePlusI>(q, input, output, sycl::plus<int>(), 0);
+  test<class KernelNameMinimumI>(q, input, output, sycl::minimum<int>(),
+                                 std::numeric_limits<int>::max());
+  test<class KernelNameMaximumI>(q, input, output, sycl::maximum<int>(),
+                                 std::numeric_limits<int>::lowest());
+
+#ifdef SPIRV_1_3
+  test<class KernelName_WonwuUVPUPOTKRKIBtT>(q, input, output,
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/reduce_sycl2020.cpp]: Please fix kernel name here as well

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 3 above)
Comment (dm-vodopyanov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: Please add `const` qualifier to variables which don't change during their lifetime. Or add `constexpr` qualifier if initialization of such variables can be moved to compile-time.

File: SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+template <typename SpecializationKernelName, typename InputContainer,
+          typename OutputContainer, class BinaryOperation>
+void test(queue q, InputContainer input, OutputContainer output,
+          BinaryOperation binary_op,
+          typename OutputContainer::value_type identity) {
+  typedef typename InputContainer::value_type InputT;
+  typedef typename OutputContainer::value_type OutputT;
+  typedef class inclusive_scan_kernel<SpecializationKernelName, 0> kernel_name0;
+  size_t N = input.size();
+  size_t G = 64;
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp]: `constexpr`?

File: SYCL/GroupAlgorithm/reduce_sycl2020.cpp [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+template <typename SpecializationKernelName, typename InputContainer,
+          typename OutputContainer, class BinaryOperation>
+void test(queue q, InputContainer input, OutputContainer output,
+          BinaryOperation binary_op,
+          typename OutputContainer::value_type identity) {
+  typedef typename InputContainer::value_type InputT;
+  typedef typename OutputContainer::value_type OutputT;
+  OutputT init = 42;
+  size_t N = input.size();
+  size_t G = 64;
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/reduce_sycl2020.cpp]: `constexpr`?

File: SYCL/GroupAlgorithm/reduce_sycl2020.cpp (hunk 10 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/reduce_sycl2020.cpp]: Fixed

File: SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp (hunk 11 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/inclusive_scan_sycl2020.cpp]: Added constexpr for N and G variables

File: SYCL/GroupAlgorithm/reduce_sycl2020.cpp (hunk 12 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/reduce_sycl2020.cpp]: Added constexpr for init, N and G variables

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 1 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:22]: It seems that Jenkins pre-ci linux/cuda compiles with c++14. Tests are failing with message "no member named 'exclusive_scan' in namespace 'std'"

Comment (dm-vodopyanov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:22]: @maximdimakov, -fsycl triggers -std=c++17, so it should be compiled with c++17. Looks strange, does it reproduce locally?

File: SYCL/GroupAlgorithm/reduce_sycl2020.cpp [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            out[0] = reduce_over_group(g, in[lid], binary_op);
+            out[1] = reduce_over_group(g, in[lid], init, binary_op);
+            out[2] = joint_reduce(g, in.get_pointer(), in.get_pointer() + N,
+                                  binary_op);
+            out[3] = joint_reduce(g, in.get_pointer(), in.get_pointer() + N,
+                                  init, binary_op);
+          });
+    });
+  }
+  // std::reduce is not implemented yet, so use std::accumulate instead
```
Comment (dm-vodopyanov) [inline SYCL/GroupAlgorithm/reduce_sycl2020.cpp:47]: Missed this during my prev review iteration. Please use `std::reduce` instead.

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 1 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:22]: I can not reproduce this problem locally. The errors occurs when I explicitly specify -std=c++14. Without -std flag or with      -std=c++17 the test passed

Comment (dm-vodopyanov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:22]: Looks like the environment problem of CI machine, could be outdated libstdc++ which doesn't contain implementation for `std::inclusive_scan` and `std::exclusive_scan`. @tfzhu, can you please take a look on this issue and try to resolve it?

File: SYCL/GroupAlgorithm/reduce_sycl2020.cpp (hunk 13 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/reduce_sycl2020.cpp:47]: Replaced

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 1 above)
Comment (dm-vodopyanov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:22]: Also tagging @bader to be aware about this CI problem.

**Comment by vladimirlaz** [comment]
@maximdimakov, could you please fix formating issue and make sure that all test are passing as soon as compiler change was merged a while ago.

**Comment by bader** [comment]
FYI: name of the standard is "SYCL 2020", not "sycl2020".

**Comment by maximdimakov** [comment]
@vladimirlaz it seems that for linux and cuda the problem remains the same

**Comment by dm-vodopyanov** [comment]
@maximdimakov, this is because of https://github.com/intel/llvm-test-suite/pull/310#discussion_r648240974. Can you please contact tools team to resolve this issue?

**Review by vladimirlaz — APPROVED** [review]

**Review by vladimirlaz — DISMISSED** [review]

**Comment by vladimirlaz** [comment]
@AlexeySachkov could you please review/approve the PR?

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// TODO: enable compile+runtime checks for operations defined in SPIR-V 1.3.
+// That requires either adding a switch to clang (-spirv-max-version=1.3) or
+// raising the spirv version from 1.1. to 1.3 for spirv translator
+// unconditionally. Using operators specific for spirv 1.3 and higher with
+// -spirv-max-version=1.1 being set by default causes assert/check fails
+// in spirv translator.
```
Comment (AlexeySachkov) [inline SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: This was already done in intel/llvm#3267 so this TODO can be removed and corresponding RUN line enabled

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp:22]: Was the problem resolved? If not, could you please leave a TODO here about using `std::exclusive_scan` later?

File: SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp (hunk 14 above)
Comment (maximdimakov) [reply SYCL/GroupAlgorithm/exclusive_scan_sycl2020.cpp]: These lines were removed

**Review by vladimirlaz — APPROVED** [review]

**Comment by vladimirlaz** [comment]
@AlexeySachkov could you please review/approve the PR?

**Review by AlexeySachkov — APPROVED** [review]

**Pull request merged**